ci: Add support for building windows on arm packages

MGN-GIT opened the pull request:
Description:
* The adoption of Windows on ARM (WoA) devices is steadily increasing, yet many applications are still not natively available for this platform.
* GitHub Actions now offer native CI runners for Windows on ARM devices (windows-11-arm), enabling automated builds and testing.
* Currently, official neovim binaries are not provided for Windows ARM64 and thus users/developers were facing difficulties using popular neovim application natively.
* This PR introduces support for building neovim application on Windows ARM64, improving accessibility for developers and end users on this emerging platform.

Changes Proposed:
- Added windows arm64 job in GitHub CI to build windows on arm neovim binaries.
- Modified CMake packaging to include packaging windows on arm binaries.
- Modified env script to install and initialize vs setup for arm64 arch.

MGN-GIT committed:
build: Add support for building win-arm64 package

MGN-GIT committed:
ci: Add support for windows arm64 VS setup

MGN-GIT committed:
ci: Add support for building windows on arm binary

MGN-GIT committed:
ci: Add support for building windows on arm binary

MGN-GIT committed:
ci: Add support for building windows on arm binary

Greenie0701 commented:
@justinmk Is there anything I am missing with this PR, not sure why this PR is not reviewed yet

justinmk commented on cmake.packaging/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
-if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64")
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(arm64|ARM64|aarch64)$")
   set(CMAKE_SYSTEM_PROCESSOR arm64)
+  set(IS_ARM64 TRUE)
```
can the conditions check CMAKE_SYSTEM_PROCESSOR instead of adding another flag?

justinmk commented:
> @justinmk Is there anything I am missing with this PR, not sure why this PR is not reviewed yet

definitely a nice improvement, just need to clean it up a bit

Greenie0701 commented on cmake.packaging/CMakeLists.txt:
```diff
@@ -27,31 +28,37 @@ set(CPACK_RESOURCE_FILE_README ${PROJECT_SOURCE_DIR}/README.md)
+  if(IS_ARM64)
+    # Windows ARM64 - only generate artifacts ZIP, since WIX is not available for Win-ARM64
+    set(CPACK_PACKAGE_FILE_NAME "nvim-winarm64")
+    set(CPACK_GENERATOR ZIP)
```
Wix installer is not natively available for windows on arm, there is only x64 version of wix

Greenie0701 replied:
But I can try finding an alternative and come up with another PR

justinmk replied:
if cmake doesn't have a builtin way to do it, let's skip it for now, if it will require a lot of new code.

Greenie0701 replied:
@justinmk actually dotnet provides wix installer irrespective of windows arch, so we that wont requires lot of new code in IMO

MGN-GIT committed:
build: make arm64/aarch64 processor detection case-insensitive

Use `(?i)` in the regex so that CMAKE_SYSTEM_PROCESSOR matches
`arm64`, `ARM64`, `aarch64`, or any case variation without
needing explicit alternates.

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

MGN-GIT committed:
cmake: make arm64/aarch64 processor detection case-insensitive

MGN-GIT committed:
ci: add support for windows arm64 neovim builds

justinmk commented:
AI lied to me about `(?i)` , can forget that :)

MGN-GIT committed:
build: fix cmake assertation for detecting arm arch

justinmk commented on .github/scripts/env.ps1:
```diff
@@ -1,9 +1,12 @@
+$installationPath = vswhere.exe -latest -requiresAny -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -requires Microsoft.VisualStudio.Component.VC.Tools.arm64 -property installationPath
+
+if ($installationPath) {
+  & "${env:COMSPEC}" /s /c "`"$installationPath\Common7\Tools\vsdevcmd.bat`" -arch=$arch -no_logo && set" |
```
Oh, I didn't realize that `matrix.arch` from the GHA yaml was feeding into this `-arch` param. I assume this won't work with `-arch=x86_64`, since the old code had `-arch=x64`.

Greenie0701 replied:
If we need x86_64 in GHA, we can have some thing like this in env.ps1:
`$arch = if ($env:BUILD_ARCH -eq "x86_64") { "x64" } else { "arm64" }
`

justinmk commented on .github/workflows/release.yml:
```diff
@@ -132,27 +132,39 @@ jobs:
-            build/nvim-win64.msi
-            build/nvim-win64.zip
+            build/${{ matrix.zip_name }}
+            ${{ matrix.arch == 'x86_64' && format('build/nvim-win64.msi') || '' }}
```
is format() needed ?

Greenie0701 replied:
yes the backslash build/ is considered as newline without format

MGN-GIT committed:
Update env.ps1

MGN-GIT committed:
build: Add support for building windows arm64 binaries

MGN-GIT committed:
ci: Add support for building windows arm64 binaries

Greenie0701 commented:
@justinmk pushed few changes to get msi installer for windows arm64 and now the changes are minimal, hope we comes to an end :))

justinmk commented on cmake.packaging/CMakeLists.txt:
```diff
@@ -27,9 +27,14 @@ set(CPACK_RESOURCE_FILE_README ${PROJECT_SOURCE_DIR}/README.md)
 
 if(WIN32)
-  set(CPACK_PACKAGE_FILE_NAME "nvim-win64")
+  if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(arm64|ARM64|aarch64)$")
```
CMAKE_SYSTEM_PROCESSOR was set on line 2

justinmk commented on .github/scripts/env.ps1:
```diff
@@ -1,9 +1,12 @@
-    $name, $value = $_ -split '=', 2
-    "$name=$value" >> $env:GITHUB_ENV
-  }
+$arch = if ($env:BUILD_ARCH -eq "x86_64") { "x64" } else { "arm64" } 
+$installationPath = vswhere.exe -latest -requiresAny -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -requires Microsoft.VisualStudio.Component.VC.Tools.arm64 -property installationPath
```
just a guess, maybe `-requiresAny` is causing problems. try something more explicit (didn't test this code, just showing the idea): 

(also, I suggest checking for "arm64" and fallback to "x64", not vice versa.

```
if ($env:BUILD_ARCH -eq "arm64") {
  $arch = "arm64"
  $installationPath = vswhere.exe -latest -requires Microsoft.VisualStudio.Component.VC.Tools.arm64 -property installationPath
} else {
  $arch = "x64"
  $installationPath = vswhere.exe -latest -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -property installationPath
}
```

justinmk commented:
> I am really surprised by seeing CI failures from files which I have not even touched :<

You touched the Windows build and Windows CI is failing to build. Focus on those CI jobs.

MGN-GIT committed:
build: Add support for packaging windows on arm artifacts

justinmk commented on .github/workflows/release.yml:
```diff
@@ -132,27 +132,44 @@ jobs:
+      - name: Install Wix
+        run: |
+          Invoke-WebRequest -Uri "https://github.com/wixtoolset/wix3/releases/download/wix3141rtm/wix314-binaries.zip" -OutFile "wix314-binaries.zip"
+          Expand-Archive -Path "wix314-binaries.zip" -DestinationPath "C:\wix"
```
you really need to be careful about yaml. I would not be surprised if `"\w"` is treated like an escape. Don't use `\` slashes unless it is absolutely required. 

`\` are rarely required on Windows, despite everyone just automatically assuming that.

justinmk replied:
my suggestion is to use "/"

MGN-GIT committed:
build: add arch-specific vswhere detection for vsdev cmd

MGN-GIT committed:
ci: Add support for building windows on arm binaries

justinmk commented on .github/workflows/release.yml:
```diff
@@ -132,27 +132,44 @@ jobs:
+    strategy:
+      matrix:
+        include:
+          - runner: windows-2022
```
future: should we change this to `windows-latest`

Greenie0701 replied:
@justinmk I believe both runners are with same configuration as could see from the runner description:
<img width="637" height="460" alt="image" src="https://github.com/user-attachments/assets/0d9c606c-202b-49fc-b912-2a9b8f2d835a" />

justinmk replied:
that will change soon, so...

Greenie0701 replied:
I have included this change as part of this PR

MGN-GIT committed:
ci: add support for building windows arm64 binaries

MGN-GIT committed:
build: Add support for packaging windows on arm artifacts

MGN-GIT committed:
ci: Use windows-latest for windows x64 builds

justinmk commented on .github/workflows/release.yml:
```diff
@@ -132,27 +132,44 @@ jobs:
+        include:
+          - runner: windows-latest
+            arch: x86_64
+            zip_name: nvim-win-x64.zip
```
don't change this for now. it affects (breaks) downstream consumers. 

the names will just be inconsistent for now. but at least the new name `nvim-win-arm64.zip` is in the "correct" format.

Greenie0701 replied:
reverted back :))

justinmk commented on .github/workflows/release.yml:
```diff
@@ -132,27 +132,44 @@ jobs:
+        uses: actions/upload-artifact@v4
         with:
-          name: nvim-win64
+          name: ${{ matrix.arch == 'x86_64' && 'nvim-win-x64' || 'nvim-win-arm64' }}
```
this name change is fine. but the zip/msi name for x64 shouldn't change.

Greenie0701 replied:
Thanks for the suggestion, made the relevant change

justinmk commented on .github/workflows/release.yml:
```diff
@@ -132,27 +132,44 @@ jobs:
+          - runner: windows-latest
+            arch: x86_64
+            zip_name: nvim-win-x64.zip
+          - runner: windows-11-arm
```
is there a "latest" variant of the windows arm runner ?

Greenie0701 replied:
The runner 'windows-11-arm' is the latest version of GHA for windows arm64, we don't have "latest" variant for windows arm64.

justinmk commented on .github/workflows/release.yml:
```diff
@@ -132,27 +132,44 @@ jobs:
+      - name: Build package
         run: |
-          cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=${{ needs.setup.outputs.build_type }}
+          cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=${{ needs.setup.outputs.build_type }} -DCMAKE_SYSTEM_PROCESSOR=${{ matrix.arch }}
```
do not pass `CMAKE_SYSTEM_PROCESSOR`, that is confusing. `CMAKE_SYSTEM_PROCESSOR` is a *cmake-provided* name. The cmake logic should check that, because that's also what users will be using when building locally. The logic should check for whatever value cmake provides on Windows x64/arm64. There is no reason to override it here, cmake already will tell us an appropriate value (and we can use `MATCHES` as already discussed).

MGN-GIT committed:
build: Add support for packaging windows arm64 binaries

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

MGN-GIT committed:
ci: Add support for building windows arm64 binaries

MGN-GIT committed:
ci: Add support for building windows arm64 binaries

justinmk commented on .github/workflows/release.yml:
```diff
@@ -132,27 +132,44 @@ jobs:
           path: |
-            build/nvim-win64.msi
-            build/nvim-win64.zip
+            build/${{ matrix.zip_name }}
+            ${{ matrix.arch == 'x86_64' && format('build/nvim-win64.msi') || format('build/nvim-win-arm64.msi') }}
```
instead of `zip_name` could we set `archive_name` and do something like this:
```suggestion
            build/${{ matrix.archive_name }}.zip
            build/${{ matrix.archive_name }}.msi
```

MGN-GIT committed:
CI: Add support for building windows arm64 binaries

justinmk commented:
I started the release workflow: https://github.com/neovim/neovim/actions/workflows/release.yml

please check that it does what you expect :) i.e. does the artifact run on windows arm?
